mempool/priority/mempool.go

```diff
@@ -0,0 +1,809 @@
+			"new_tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+			"new_priority", priority,
+		)
+
+		// Sort lowest priority items first so they will be evicted first.  Break
+		// ties in favor of newer items (to maintain FIFO semantics in a group).
+		sort.Slice(victims, func(i, j int) bool {
+			iw := victims[i].Value.(*WrappedTx)
+			jw := victims[j].Value.(*WrappedTx)
+			if iw.Priority() == jw.Priority() {
+				return iw.timestamp.After(jw.timestamp)
+			}
+			return iw.Priority() < jw.Priority()
+		})
+
+		// Evict as many of the victims as necessary to make room.
+		var evictedBytes int64
+		for _, vic := range victims {
+			w := vic.Value.(*WrappedTx)
+
+			txmp.logger.Debug(
+				"evicted valid existing transaction; mempool full",
+				"old_tx", fmt.Sprintf("%X", w.tx.Hash()),
+				"old_priority", w.priority,
+			)
+			txmp.removeTxByElement(vic)
+			txmp.cache.Remove(w.tx)
+			txmp.metrics.EvictedTxs.Add(1)
+			// Add it to evicted transactions cache
+			txmp.evictedTxs.Push(w.tx)
+			// We may not need to evict all the eligible transactions.  Bail out
+			// early if we have made enough room.
+			evictedBytes += w.Size()
+			if evictedBytes >= wtx.Size() {
+				break
+			}
+		}
+	}
+
+	wtx.SetGasWanted(checkTxRes.GasWanted)
+	wtx.SetPriority(priority)
+	wtx.SetSender(string(sender))
+	txmp.insertTx(wtx)
+
+	txmp.metrics.TxSizeBytes.Observe(float64(wtx.Size()))
+	txmp.metrics.Size.Set(float64(txmp.Size()))
+	txmp.metrics.SizeBytes.Set(float64(txmp.SizeBytes()))
+	txmp.logger.Debug(
+		"inserted new valid transaction",
+		"priority", wtx.Priority(),
+		"tx", fmt.Sprintf("%X", wtx.tx.Hash()),
+		"height", txmp.height,
+		"num_txs", txmp.Size(),
+	)
+	txmp.notifyTxsAvailable()
+}
+
+func (txmp *TxMempool) insertTx(wtx *WrappedTx) {
+	elt := txmp.txs.PushBack(wtx)
+	txmp.txByKey[wtx.tx.Key()] = elt
+	if s := wtx.Sender(); s != "" {
+		txmp.txBySender[s] = elt
+	}
+
+	atomic.AddInt64(&txmp.txsBytes, wtx.Size())
+}
+
+// handleRecheckResult handles the responses from ABCI CheckTx calls issued
+// during the recheck phase of a block Update.  It removes any transactions
+// invalidated by the application.
+//
+// This method is NOT executed for the initial CheckTx on a new transaction;
+// that case is handled by addNewTransaction instead.
+func (txmp *TxMempool) handleRecheckResult(tx types.Tx, checkTxRes *abci.ResponseCheckTx) {
+	txmp.metrics.RecheckTimes.Add(1)
+
+	// Find the transaction reported by the ABCI callback. It is possible the
+	// transaction was evicted during the recheck, in which case the transaction
+	// will be gone.
+	elt, ok := txmp.txByKey[tx.Key()]
+	if !ok {
+		return
+	}
+	wtx := elt.Value.(*WrappedTx)
+
+	// If a postcheck hook is defined, call it before checking the result.
+	var err error
+	if txmp.postCheckFn != nil {
+		err = txmp.postCheckFn(tx, checkTxRes)
+	}
+
+	if checkTxRes.Code == abci.CodeTypeOK && err == nil {
+		priority := getTxPriority(int64(checkTxRes.GasFees), checkTxRes.GasWanted)
```

Comment:
It looks like we'll never need to update the priority but good to have this in place just in case